Replaces usages of `RestStatus` with the more detailed `HttpErrorStatus`

seut opened the pull request:
RestStatus was not used anymore to deliver a HTTP error code to the client, it was supersede by the more detailed HttpErrorStatus.
This commit adds concrete `HttpErrorStatus` values to all exceptions which had a concrete `RestStatus` before.
The `RestStatus` is still needed for BWC (it may be streamed by older nodes in mixed clusters), such it's not removed but marked as deprecated instead.

Additionally, the first commit fixes some wrong detailed http error codes for existing errors.

seut commented:
retest this please

Unrelated blackbox jmx error (flaky?)

mfussenegger commented on server/src/main/java/org/elasticsearch/cluster/block/ClusterBlock.java:
```diff
@@ -47,17 +48,28 @@ public ClusterBlock(StreamInput in) throws IOException {
         retryable = in.readBoolean();
         disableStatePersistence = in.readBoolean();
-        status = RestStatus.readFrom(in);
+        status = HttpErrorStatus.readFrom(in);
```
Is this safe to replace like this?

seut replied:
No, it isn't. Thanks for catching.

mfussenegger commented on server/src/main/java/org/elasticsearch/common/io/stream/NotSerializableExceptionWrapper.java:
```diff
@@ -36,7 +37,7 @@
 
     private final String name;
-    private final RestStatus status;
+    private final HttpErrorStatus status;
```
I wonder if we even need this. Looking at the usages of `status()` there's one use in `DefaultShardOperationFailedException` which this PR removes and another in `SQLExceptions.status` - which all look like they could be removed too. 

Maybe https://github.com/crate/crate/blob/f3ff8fe20edc47a7b452e33da68e96a718213636/server/src/main/java/org/elasticsearch/index/shard/IndexShard.java#L1450 is a bit trickier to replace(?)

seut replied:
Yes, at `master`, the default `ElasticsearchException.httpErrorStatus()` impl.  (`UNHANDLED_SERVER_ERROR`) is used for the `NotSerializableException`. But given that we want to move to `HttpErrorStatus` in general everywhere, I'm not sure I understand your point correctly about that this isn't needed. Can you please clarify?

mfussenegger replied:
> But given that we want to move to HttpErrorStatus in general everywhere, I'm not sure I understand your point correctly about that this isn't needed. Can you please clarify?

My thinking was that if the exception has a specific http status code we should probably make it an `ElasticsearchException` that's streamable, so there wouldn't be a need for a `status` field in `NotSerializableExceptionWrapper` because it would always end up in the default `return RestStatus.INTERNAL_SERVER_ERROR;` case of `SQLExceptions.status()` anyway

mfussenegger commented on server/src/main/java/io/crate/replication/logical/action/ReplayChangesAction.java:
```diff
@@ -163,10 +162,7 @@ void performOnPrimary(IndexShard primary,
-                    (engineResult.getResultType() == Engine.Result.Type.FAILURE &&
-                     engineResult.getFailure() instanceof StrictDynamicMappingException)
-                ) {
+                if (engineResult.getResultType() == Engine.Result.Type.MAPPING_UPDATE_REQUIRED) {
```
Just to verify I understand this right: This is because `StrictDynamicMappingException` was never used and is removed, making the old condition here dead code?

seut replied:
Yes, exactly.

mfussenegger reviewed (approved):
Overall looks good, left a couple of minor comments

seut commented on server/src/main/java/io/crate/rest/action/HttpErrorStatus.java:
```diff
@@ -57,20 +101,49 @@ public enum HttpErrorStatus {
+
+    static {
+        HttpErrorStatus[] values = values();
+        Map<Integer, HttpErrorStatus> codeToStatus = new HashMap<>(values.length);
```
I think this is still needed as the `ClusterBlocks` requires it for BWC even if we get rid of the usage at the `NotSerializableExceptionWrapper`.

seut commented:
@mfussenegger
I've rebased the PR with the recent master and added another fixup commit to address the outstanding comments. Please check again, thx!
